### ISSUE TITLE
Best way to implement Redis caching

### ISSUE THREAD

**Issue by chriso0710** [typhoeus/typhoeus#545]
Hi all,

I could not find any info about caching Typhoeus requests with Redis.
So I am trying to figure out the best method to do this. Below is what I have got so far and which seems to work.

Is there a better way or am I missing something? Any thoughts on this?

Christian

```ruby
require "oj"

class Cache
    def initialize(ttl = 3600)
        @store = Redis.new(:url => ENV['REDIS'])
        @ttl = ttl
    end

    def get(request)
        entry = @store.get(request.cache_key)
        response = nil
        if entry then
            puts "get cached #{request.cache_key}"
            j = Oj.load entry
            response = Typhoeus::Response.new({
                :code =>        j["code"],
                :headers =>     j["headers"],
                :body =>        j["body"],
                :return_code => j["return_code"].to_sym
              })
            response.instance_variable_set :@cached, true
        end
        response
    end

    def set(request, response)
        entry = Oj.dump( { "code" => response.code, "headers" => response.headers, "body" => response.body, "return_code" => response.return_code.to_s }, :mode => :strict )
        puts "set #{request.cache_key}: #{entry}"
        @store.set(request.cache_key, entry)
        @store.expire(request.cache_key, @ttl)
    end
end
```

```ruby
2.2.1 :018 > Typhoeus::Config.cache = Cache.new
2.2.1 :019 > Typhoeus.get("www.example.com").cached?
set 79add1cc744c0511f7ce73c615e47db3b5576e0b: {"code":200,"headers":{"Cache-Control":"max-age=604800","Content-Type":"text/html","Date":"Thu, 03 Nov 2016 11:26:42 GMT","Etag":"\"359670651+gzip+ident\"","Expires":"Thu, 10 Nov 2016 11:26:42 GMT","Last-Modified":"Fri, 09 Aug 2013 23:54:35 GMT","Server":"ECS (iad/182A)","Vary":"Accept-Encoding","X-Cache":"HIT","x-ec-custom-error":"1","Content-Length":"1270"},"body":"<!doctype html>\n<html>\n<head>\n    <title>Example Domain</title>\n\n    <meta charset=\"utf-8\" />\n    <meta http-equiv=\"Content-type\" content=\"text/html; charset=utf-8\" />\n    <meta name=\"viewport\" content=\"width=device-width, initial-scale=1\" />\n    <style type=\"text/css\">\n    body {\n        background-color: #f0f0f2;\n        margin: 0;\n        padding: 0;\n        font-family: \"Open Sans\", \"Helvetica Neue\", Helvetica, Arial, sans-serif;\n        \n    }\n    div {\n        width: 600px;\n        margin: 5em auto;\n        padding: 50px;\n        background-color: #fff;\n        border-radius: 1em;\n    }\n    a:link, a:visited {\n        color: #38488f;\n        text-decoration: none;\n    }\n    @media (max-width: 700px) {\n        body {\n            background-color: #fff;\n        }\n        div {\n            width: auto;\n            margin: 0 auto;\n            border-radius: 0;\n            padding: 1em;\n        }\n    }\n    </style>    \n</head>\n\n<body>\n<div>\n    <h1>Example Domain</h1>\n    <p>This domain is established to be used for illustrative examples in documents. You may use this\n    domain in examples without prior coordination or asking for permission.</p>\n    <p><a href=\"http://www.iana.org/domains/example\">More information...</a></p>\n</div>\n</body>\n</html>\n","return_code":"ok"}
false
2.2.1 :020 > Typhoeus.get("www.example.com").cached?
get cached 79add1cc744c0511f7ce73c615e47db3b5576e0b
true
```

**Comment by i0rek** [comment]
Hello @chriso0710, thanks for writing. Your implementation seems fine. We could add it to the readme so that others can use it as well. Would you like to create a PR? Thanks again :)

**Comment by chriso0710** [comment]
Thanks for your friendly response. 
Yes, I can do that. Might take a few days though...

**Comment by isaacseymour** [comment]
@i0rek would you be open to having some common cache adapters (Redis, Rails.cache, Dalli) included in Typhoeus itself, or do you think they belong in a separate gem? Either way I'd be happy to write some :)

**Comment by i0rek** [comment]
Yes, we could do that. Lets see how it goes.

**Comment by isaacseymour** [comment]
I think this issue can be closed now that #546 is merged 😄

**Comment by i0rek** [comment]
Thats right!
